xdsl/transforms/experimental/ConvertStencilToLLMLIR.py

```diff
@@ -246,9 +261,60 @@ def match_and_rewrite(self, op: ApplyOp, rewriter: PatternRewriter, /):
             body=current_region,
         )
 
+        # Handle returnd values
+        new_results: list[SSAValue] = []
+        for result in op.res:
+            assert isa(
+                result.typ, TempType[Attribute]
+            ), f"Expected return value to be a !{TempType.name}"
+            assert isinstance(
+                result.typ.bounds, StencilBoundsAttr
+            ), f"Expected output to be sized before lowering. {result.typ}"
+            shape = result.typ.get_shape()
+            element_type = result.typ.element_type
+
+            # If it is buffered, allocate the buffer
+            if any(isinstance(use.operation, BufferOp) for use in result.uses):
+                alloc = memref.Alloc.get(element_type, shape=shape)
+                alloc_type = alloc.memref.typ
+                assert isa(alloc_type, MemRefType[Attribute])
+
+                offset = list(-result.typ.bounds.lb)
+
+                view = memref.Subview.from_static_parameters(
+                    alloc,
+                    alloc_type,
+                    offset,
+                    shape,
+                    [1] * result.typ.get_num_dims(),
+                )
+                rewriter.insert_op_before_matched_op((alloc, view))
+                new_results.append(view.result)
+                update_return_target(self.return_targets, result, view.result)
+            else:
+                new_results.append(result)
+
+        deallocs: list[Operation] = []
+        # Handle input buffer deallocation
+        for input in op.args:
+            # Is this input a temp buffer?
+            if isinstance(input.typ, TempType) and isinstance(input.owner, BufferOp):
+                block = op.parent_block()
+                assert block is not None
+                self_index = block.get_operation_index(op)
+                # Is it its last use?
+                if not any(
+                    use.operation.parent_block() is block
+                    and block.get_operation_index(use.operation) > self_index
+                    for use in input.uses
+                ):
+                    # Then deallocate it
+                    deallocs.append(memref.Dealloc.get(input))
+
         # Replace with the loop and necessary constants.
         rewriter.insert_op_before_matched_op([*lowerBounds, one, *upperBounds, p])
-        rewriter.erase_matched_op()
+        rewriter.insert_op_after_matched_op([*deallocs])
+        rewriter.replace_matched_op([], new_results)
```

Comment:
This part is ugly; it should be just one replace_matched_op.
Though doing so renames a looot of variables in the filechecks, putting again a lot of noise on checking the functionality. So I'm postponing that!